File: frontend/src/services/api/capacity.ts
/**
 * Capacity APIs - Team capacity settings, health checks, and CFD.
 */
import api from './client';
import type {
  CapacitySetting,
  TeamCapacitySummary,
  TeamHealthCheck,
  TeamHealthSummary,
  CFDData,
  FlowMetrics,
  PaginatedResponse,
} from './types';
import { DEFAULT_PAGE_SIZE } from './types';
import { normalizePaginatedResponse } from './pagination';

// =============================================================================
// Capacity Settings CRUD
// =============================================================================

export const listCapacitySettings = async (opts?: {
  projectId?: number;
  assigneeEmail?: string;
  includeExpired?: boolean;
  skip?: number;
  limit?: number;
}): Promise<PaginatedResponse<CapacitySetting>> => {
  const params: Record<string, unknown> = {};
  if (opts?.projectId !== undefined) params.project_id = opts.projectId;
  if (opts?.assigneeEmail) params.assignee_email = opts.assigneeEmail;
  if (opts?.includeExpired) params.include_expired = opts.includeExpired;
  const skip = opts?.skip ?? 0;
  const limit = opts?.limit ?? DEFAULT_PAGE_SIZE;
  params.skip = skip;
  params.limit = limit;

  const { data } = await api.get('/v1/capacity/settings', { params });
  return normalizePaginatedResponse<CapacitySetting>(data, { skip, limit });
};

export const getCapacitySetting = async (settingId: number): Promise<CapacitySetting> => {
  const { data } = await api.get(`/v1/capacity/settings/${settingId}`);
  return data as CapacitySetting;
};

export const createCapacitySetting = async (setting: {
  projectId?: number;
  assigneeEmail: string;
  assigneeName?: string;
  hoursPerWeek: number;
  focusFactor?: number;
  validFrom?: string;
  validTo?: string;
  notes?: string;
}): Promise<CapacitySetting> => {
  const body: Record<string, unknown> = {
    assignee_email: setting.assigneeEmail,
    hours_per_week: setting.hoursPerWeek,
  };
  if (setting.projectId !== undefined) body.project_id = setting.projectId;
  if (setting.assigneeName !== undefined) body.assignee_name = setting.assigneeName;
  if (setting.focusFactor !== undefined) body.focus_factor = setting.focusFactor;
  if (setting.validFrom !== undefined) body.valid_from = setting.validFrom;
  if (setting.validTo !== undefined) body.valid_to = setting.validTo;
  if (setting.notes !== undefined) body.notes = setting.notes;

  const { data } = await api.post('/v1/capacity/settings', body);
  return data as CapacitySetting;
};

export const updateCapacitySetting = async (
  settingId: number,
  update: Partial<{
    projectId: number;
    assigneeEmail: string;
    assigneeName: string;
    hoursPerWeek: number;
    focusFactor: number;
    validFrom: string;
    validTo: string;
    notes: string;
  }>
): Promise<CapacitySetting> => {
  const body: Record<string, unknown> = {};
  if (update.projectId !== undefined) body.project_id = update.projectId;
  if (update.assigneeEmail !== undefined) body.assignee_email = update.assigneeEmail;
  if (update.assigneeName !== undefined) body.assignee_name = update.assigneeName;
  if (update.hoursPerWeek !== undefined) body.hours_per_week = update.hoursPerWeek;
  if (update.focusFactor !== undefined) body.focus_factor = update.focusFactor;
  if (update.validFrom !== undefined) body.valid_from = update.validFrom;
  if (update.validTo !== undefined) body.valid_to = update.validTo;
  if (update.notes !== undefined) body.notes = update.notes;

  const { data } = await api.patch(`/v1/capacity/settings/${settingId}`, body);
  return data as CapacitySetting;
};

export const deleteCapacitySetting = async (settingId: number): Promise<void> => {
  await api.delete(`/v1/capacity/settings/${settingId}`);
};

// =============================================================================
// Team Capacity Summary
// =============================================================================

export const getTeamCapacitySummary = async (
  projectId: number,
  opts?: { sprintWeeks?: number; referenceDate?: string }
): Promise<TeamCapacitySummary> => {
  const params: Record<string, unknown> = { project_id: projectId };
  if (opts?.sprintWeeks !== undefined) params.sprint_weeks = opts.sprintWeeks;
  if (opts?.referenceDate) params.reference_date = opts.referenceDate;

  const { data } = await api.get('/v1/capacity/summary', { params });
  return data as TeamCapacitySummary;
};

// =============================================================================
// Team Health Checks
// =============================================================================

export const listTeamHealthChecks = async (
  projectId: number,
  opts?: { sprintId?: number; skip?: number; limit?: number }
): Promise<PaginatedResponse<TeamHealthCheck>> => {
  const params: Record<string, unknown> = { project_id: projectId };
  if (opts?.sprintId) params.sprint_id = opts.sprintId;
  const skip = opts?.skip ?? 0;
  const limit = opts?.limit ?? DEFAULT_PAGE_SIZE;
  params.skip = skip;
  params.limit = limit;

  const { data } = await api.get('/v1/capacity/health-checks', { params });
  return normalizePaginatedResponse<TeamHealthCheck>(data, { skip, limit });
};

export const createTeamHealthCheck = async (check: {
  projectId: number;
  sprintId?: number;
  satisfaction?: number;
  workloadBalance?: number;
  technicalDebtPressure?: number;
  collaborationQuality?: number;
  happinessIndex?: number;
  burnoutRiskScore?: number;
  burnoutRiskFactors?: string;
  checkDate: string;
  respondentCount?: number;
  notes?: string;
}): Promise<TeamHealthCheck> => {
  const body: Record<string, unknown> = {
    project_id: check.projectId,
    check_date: check.checkDate,
  };
  if (check.sprintId !== undefined) body.sprint_id = check.sprintId;
  if (check.satisfaction !== undefined) body.satisfaction = check.satisfaction;
  if (check.workloadBalance !== undefined) body.workload_balance = check.workloadBalance;
  if (check.technicalDebtPressure !== undefined) body.technical_debt_pressure = check.technicalDebtPressure;
  if (check.collaborationQuality !== undefined) body.collaboration_quality = check.collaborationQuality;
  if (check.happinessIndex !== undefined) body.happiness_index = check.happinessIndex;
  if (check.burnoutRiskScore !== undefined) body.burnout_risk_score = check.burnoutRiskScore;
  if (check.burnoutRiskFactors !== undefined) body.burnout_risk_factors = check.burnoutRiskFactors;
  if (check.respondentCount !== undefined) body.respondent_count = check.respondentCount;
  if (check.notes !== undefined) body.notes = check.notes;

  const { data } = await api.post('/v1/capacity/health-checks', body);
  return data as TeamHealthCheck;
};

export const getTeamHealthSummary = async (
  projectId: number,
  opts?: { periodDays?: number }
): Promise<TeamHealthSummary> => {
  const params: Record<string, unknown> = { project_id: projectId };
  if (opts?.periodDays) params.period_days = opts.periodDays;

  const { data } = await api.get('/v1/capacity/health-summary', { params });
  return data as TeamHealthSummary;
};

// =============================================================================
// CFD (Cumulative Flow Diagram)
// =============================================================================

export const getCFDData = async (
  projectId: number,
  opts?: { sprintId?: number; startDate?: string; endDate?: string; limit?: number }
): Promise<CFDData> => {
  const params: Record<string, unknown> = { project_id: projectId };
  if (opts?.sprintId) params.sprint_id = opts.sprintId;
  if (opts?.startDate) params.start_date = opts.startDate;
  if (opts?.endDate) params.end_date = opts.endDate;
  if (opts?.limit !== undefined) params.limit = opts.limit;

  const { data } = await api.get('/v1/capacity/cfd', { params });
  return data as CFDData;
};

export const getFlowMetrics = async (
  projectId: number,
  opts?: { sprintId?: number; days?: number }
): Promise<FlowMetrics> => {
  const params: Record<string, unknown> = { project_id: projectId };
  if (opts?.sprintId) params.sprint_id = opts.sprintId;
  if (opts?.days) params.days = opts.days;

  const { data } = await api.get('/v1/capacity/flow-metrics', { params });
  return data as FlowMetrics;
};

export const takeCFDSnapshot = async (
  projectId: number,
  opts?: { sprintId?: number }
): Promise<{ message: string; snapshot_id: number }> => {
  const params: Record<string, unknown> = { project_id: projectId };
  if (opts?.sprintId) params.sprint_id = opts.sprintId;

  const { data } = await api.post('/v1/capacity/cfd/snapshot', undefined, { params });
  return data;
};

File: backend/app/schemas/capacity.py
"""
Pydantic schemas for capacity settings, team health checks, and CFD snapshots.
"""

from typing import Optional, List, Dict, Any
from datetime import date, datetime
from pydantic import BaseModel, ConfigDict, Field, field_validator


# =============================================================================
# Capacity Settings Schemas
# =============================================================================


class CapacitySettingsBase(BaseModel):
    """Base schema for capacity settings."""

    assignee_email: str = Field(..., max_length=255)
    assignee_name: Optional[str] = Field(None, max_length=255)
    hours_per_week: float = Field(default=40.0, ge=0, le=168)  # Max 168h = 24*7
    focus_factor: float = Field(default=0.8, ge=0.0, le=1.0)
    valid_from: Optional[date] = None
    valid_to: Optional[date] = None
    notes: Optional[str] = Field(None, max_length=500)

    @field_validator("focus_factor")
    @classmethod
    def validate_focus_factor(cls, v):
        """Ensure focus factor is a valid percentage."""
        if v < 0 or v > 1:
            raise ValueError("Focus factor must be between 0 and 1")
        return v


class CapacitySettingsCreate(CapacitySettingsBase):
    """Schema for creating capacity settings."""

    project_id: Optional[int] = None


class CapacitySettingsUpdate(BaseModel):
    """Schema for updating capacity settings."""

    assignee_name: Optional[str] = Field(None, max_length=255)
    hours_per_week: Optional[float] = Field(None, ge=0, le=168)
    focus_factor: Optional[float] = Field(None, ge=0.0, le=1.0)
    valid_from: Optional[date] = None
    valid_to: Optional[date] = None
    notes: Optional[str] = Field(None, max_length=500)


class CapacitySettingsInDB(CapacitySettingsBase):
    """Schema for capacity settings from database."""

    id: int
    project_id: Optional[int] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class CapacitySettings(CapacitySettingsInDB):
    """Full capacity settings response schema."""

    effective_capacity: Optional[float] = None

    @classmethod
    def from_orm_with_effective(cls, obj):
        """Create instance with calculated effective capacity."""
        data = {
            "id": obj.id,
            "project_id": obj.project_id,
            "assignee_email": obj.assignee_email,
            "assignee_name": obj.assignee_name,
            "hours_per_week": obj.hours_per_week,
            "focus_factor": obj.focus_factor,
            "valid_from": obj.valid_from,
            "valid_to": obj.valid_to,
            "notes": obj.notes,
            "created_at": obj.created_at,
            "updated_at": obj.updated_at,
            "effective_capacity": obj.effective_capacity,
        }
        return cls(**data)


class TeamCapacitySummary(BaseModel):
    """Summary of team capacity for a sprint."""

    total_theoretical_hours: float
    total_effective_hours: float
    average_focus_factor: float
    team_members: int
    capacity_by_member: List[Dict[str, Any]]


# =============================================================================
# Team Health Check Schemas
# =============================================================================


class TeamHealthCheckBase(BaseModel):
    """Base schema for team health checks."""

    satisfaction: Optional[float] = Field(None, ge=1, le=5)
    workload_balance: Optional[float] = Field(None, ge=1, le=5)
    technical_debt_pressure: Optional[float] = Field(None, ge=1, le=5)
    collaboration_quality: Optional[float] = Field(None, ge=1, le=5)
    happiness_index: Optional[float] = Field(None, ge=1, le=5)
    burnout_risk_score: Optional[float] = Field(None, ge=0, le=1)
    burnout_risk_factors: Optional[str] = Field(None, max_length=500)
    check_date: date
    respondent_count: Optional[int] = Field(None, ge=0)
    notes: Optional[str] = Field(None, max_length=1000)


class TeamHealthCheckCreate(TeamHealthCheckBase):
    """Schema for creating team health check."""

    project_id: int
    sprint_id: Optional[int] = None


class TeamHealthCheckUpdate(BaseModel):
    """Schema for updating team health check."""

    satisfaction: Optional[float] = Field(None, ge=1, le=5)
    workload_balance: Optional[float] = Field(None, ge=1, le=5)
    technical_debt_pressure: Optional[float] = Field(None, ge=1, le=5)
    collaboration_quality: Optional[float] = Field(None, ge=1, le=5)
    happiness_index: Optional[float] = Field(None, ge=1, le=5)
    burnout_risk_score: Optional[float] = Field(None, ge=0, le=1)
    burnout_risk_factors: Optional[str] = Field(None, max_length=500)
    respondent_count: Optional[int] = Field(None, ge=0)
    notes: Optional[str] = Field(None, max_length=1000)


class TeamHealthCheckInDB(TeamHealthCheckBase):
    """Schema for team health check from database."""

    id: int
    project_id: int
    sprint_id: Optional[int] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TeamHealthCheck(TeamHealthCheckInDB):
    """Full team health check response schema."""

    pass


class HealthTrend(BaseModel):
    """Health metrics trend over time."""

    check_date: date
    happiness_index: Optional[float]
    burnout_risk_score: Optional[float]


class TeamHealthSummary(BaseModel):
    """Summary of team health metrics."""

    latest_happiness_index: Optional[float]
    latest_burnout_risk: Optional[float]
    trend_direction: str  # "improving", "declining", "stable"
    checks_count: int
    trend_data: List[HealthTrend]


# =============================================================================
# CFD Snapshot Schemas
# =============================================================================


class CFDSnapshotBase(BaseModel):
    """Base schema for CFD snapshots."""

    snapshot_date: date
    backlog_count: int = 0
    todo_count: int = 0
    in_progress_count: int = 0
    in_review_count: int = 0
    testing_count: int = 0
    done_count: int = 0
    total_count: int = 0
    wip_count: int = 0
    throughput: Optional[int] = None
    avg_cycle_time_hours: Optional[float] = None


class CFDSnapshotCreate(CFDSnapshotBase):
    """Schema for creating CFD snapshot."""

    project_id: int
    sprint_id: Optional[int] = None


class CFDSnapshotInDB(CFDSnapshotBase):
    """Schema for CFD snapshot from database."""

    id: int
    project_id: int
    sprint_id: Optional[int] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CFDSnapshot(CFDSnapshotInDB):
    """Full CFD snapshot response schema."""

    pass


class CFDData(BaseModel):
    """CFD data for visualization."""

    snapshots: List[CFDSnapshot]
    date_range: Dict[str, date]
    status_labels: List[str]


class FlowMetrics(BaseModel):
    """Flow metrics derived from CFD data."""

    avg_lead_time_days: Optional[float]
    avg_cycle_time_hours: Optional[float]
    avg_throughput_per_day: Optional[float]
    wip_trend: str  # "increasing", "decreasing", "stable"
    bottleneck_status: Optional[str]  # Status with highest accumulation
